codebuild/bin/install_default_dependencies.sh

```diff
@@ -37,6 +37,14 @@ if [[ ("$S2N_LIBCRYPTO" == "openssl-1.1.1") || ("$TESTS" == "integration" || "$T
     fi
 fi
 
+# Download and Install Openssl 3.0
+if [[ ("$S2N_LIBCRYPTO" == "openssl-3") || ("$TESTS" == "integration" || "$TESTS" == "integrationv2" || "$TESTS" == "ALL" ) ]]; then
```

Comment:
Why limit this to specific tests? Don't you need it to build s2n-tls regardless of what tests are being run?

Reply:
sort of a historical pattern; really there should be only two cases: explicitly include this library with `S2N_LIBCRYPTO` and `ALL`.  Our current approach where we want to break out individual tests with specific libcryptos can be sped up by not installing everything, but hopefully the custom docker container work will move all of the install steps to container build time.

Reply:
I took the code bit from openssl 1.1.1 example. But if i am understanding this correctly, the default is openssl1.1.1 and for all other libcrypto we rebuild?

I am actually not sure why we gate around certain test for openssl 1.1.1 above(unit test seems to be missing).

Reply:
> sort of a historical pattern; really there should be only two cases: explicitly include this library with S2N_LIBCRYPTO and ALL. Our current approach where we want to break out individual tests with specific libcryptos can be sped up by not installing everything, but hopefully the custom docker container work will move all of the install steps to container build time.

Isn't that kind of an accident tho? It just happens that we only test multiple different libcryptos in our integration test CIs. We SHOULDN'T be running any other test in the CI with S2N_LIBCRYPTO set to something else. But if we don't download the right libcrypto if we do end up doing that, it seems like a bug. And people could use this locally too.

Reply:
If nothing else, it seems like the build should fail if an unexpected libcrypto is requested. We don't just silently fall back to building with the system libcrypto or openssl-1.1.1, right?

Reply:
Not super familiar with the entire build process but it does seem odd and error-prone that we install openssl-1.1.1 even though its not specified.

From local testing the linker seems to do eager search and does link to the a different openssl version if the original one is not there. While this may not be an issue, it might be safer to only build an openssl version if we are testing with it.. custom docker images with only one version available sounds like the better option.

Reply:
That sounds like we could definitely end up incorrectly thinking we're testing against libcryptos we're not actually testing against :\ Can you create an issue for that?

Reply:
https://github.com/aws/s2n-tls/issues/3455